### PR TITLE
NAS-111288 / 21.08 / Forward port enclosure changes to scale

### PR DESCRIPTION
**Pull request by dszidi**
There were quite a few bug fixes as well as R series support that needed to be ported to scale.
This includes all the visualization files from src/app/core/classes/hardware as well as all the type safety and linter fixes

This PR should be tested against an R Series and a supported MINI.

Testing Criteria:
- Compiler and linter should not complain
- Users should be able to click on drive slots in the visualization to select disks
- Hitting the Identify button should flash the lights on the hardware drive slot (Not available on MINIs)
- The "Elements" button in the page title  should allow users to switch between views (Not available on MINIs)
- The "Edit Label" button should allow users to update the enclosure name that appears on the card title (Not available on MINIs)
- The Disks view should update without refreshing when adding or removing a disk from the unit. (Requires physical access)

### CODE REVIEW

**Comment by undsoft** [comment]
Let's pick one of the enclosure variants and generally test that it works. Specific models can be tested by RE as needed.

**Comment by dszidi** [comment]
> Let's pick one of the enclosure variants and generally test that it works. Specific models can be tested by RE as needed.

Ok. In that case we should do the two that were specifically mentioned in the ticket. R Series and a supported MINI. Will update testing instructionse

**Comment by AlexKarpov98** [comment]
@dszidi There is a UI layout issue, the bottom of page is not shown  correctly.
image
image

**Comment by AlexKarpov98** [comment]
@dszidi It shouldn’t have that button if there is only one menu item
image

**Comment by dszidi** [comment]
> @dszidi Is it ok not to show the information that is shown in 12.0?...
Not for this ticket. We just want to forward port changes for right now.

File: src/app/core/classes/hardware/chassis-view.ts [hunk 1]
```diff
@@ -50,14 +64,18 @@ export class ChassisView {
...
   columns: number;
+  orientation = 'rows'; // 'rows' || 'columns'
+  layout?: LayoutGenerator;
   vertical = false;
   filters: any[] = [];
   disabledOpacity = 0.25;
   chassisOpacity = 0.25;
   initialized = false;
-  loader: PIXI.loaders.Loader;
+  loader: any;
```
Comment (undsoft) [inline src/app/core/classes/hardware/chassis-view.ts:74]: Any reason to for `any` here?

File: src/app/core/classes/hardware/chassis-view.ts [hunk 2]
```diff
@@ -81,7 +99,7 @@ export class ChassisView {
 
   requiredAssets(): { alias: string; path: string }[] {
     // Return a list of assets for the loader to fetch
-    const assets: { alias: string; path: string }[] = [];
+    const assets: any[] = [];
```
Comment (undsoft) [inline src/app/core/classes/hardware/chassis-view.ts:102]: Reducing type safety seems to be unnecessary.

File: src/app/core/classes/hardware/chassis-view.ts [hunk 3]
```diff
@@ -129,11 +147,21 @@ export class ChassisView {
...
+
     this.chassis.name = this.model + '_chassis';
     this.chassis.alpha = 0;
-    this.chassis.scale.x = this.chassisScale && this.chassisScale.x ? this.chassisScale.x : 1;
-    this.chassis.scale.y = this.chassisScale && this.chassisScale.y ? this.chassisScale.y : 1;
+    this.chassis.x = this.chassisOffsetX;
+    this.chassis.y = this.chassisOffsetY;
+
+    this.chassis.scale.x = this.chassisScale && this.chassisScale.x
+      ? this.chassisScale.x : 1;
```
Comment (undsoft) [inline src/app/core/classes/hardware/chassis-view.ts:160]: Either single line X ? Y : Z or
```
X
  ? Y
  : Z
```

File: src/app/core/classes/hardware/vdev-labels-svg.ts [hunk 4]
```diff
@@ -43,10 +44,11 @@ export class VDevLabelsSVG {
     this.app = app;
     this.mainStage = this.app.stage;
     this.d3Init();
+    // const paths = this.getParent().querySelectorAll('svg path');
```
Comment (undsoft) [inline src/app/core/classes/hardware/vdev-labels-svg.ts:47]: If not needed, then nuke it.

File: src/app/core/classes/system-profiler.ts [hunk 5]
```diff
@@ -45,23 +44,23 @@ export class SystemProfiler {
     this._diskData = null;
     this._diskData = obj;
     this.parseDiskData(obj);
-    this.parseEnclosures();
+    this.parseEnclosures(/* this._enclosures */);
```
Comment (undsoft) [inline src/app/core/classes/system-profiler.ts:47]: What does comment do?

File: src/app/core/classes/system-profiler.ts [hunk 6]
```diff
@@ -291,8 +289,8 @@ export class SystemProfiler {
 
   getEnclosureNumber(diskName: string): number {
     // To be deprecated when middleware includes enclosure number with disk info
-    let result: number;
-    this.profile.forEach((enclosure: EnclosureMetadata, index: number) => {
+    let result;
```
Comment (undsoft) [inline src/app/core/classes/system-profiler.ts:292]: While types on `forEach` are unnecessary, removing `number` from `number` slightly reduces its typesafety until first assignment.

File: src/app/pages/system/view-enclosure/enclosure-disks/enclosure-disks.component.ts [hunk 7]
```diff
@@ -117,7 +122,9 @@ export class EnclosureDisksComponent implements AfterContentInit, OnChanges, OnD
   }
 
   get enclosurePools(): any {
-    return Object.keys(this.subenclosure ? this.subenclosure.poolKeys : this.selectedEnclosure.poolKeys);
+    const selectedEnclosure = this.getSelectedEnclosure();
+    return Object.keys(selectedEnclosure.poolKeys);
+    // return Object.keys(this.subenclosure ? this.subenclosure.poolKeys : this.selectedEnclosure.poolKeys);
```
Comment (undsoft) [inline src/app/pages/system/view-enclosure/enclosure-disks/enclosure-disks.component.ts:127]: Unused.

File: src/app/pages/system/view-enclosure/view-enclosure.component.ts [hunk 8]
```diff
@@ -241,5 +259,37 @@ export class ViewEnclosureComponent implements AfterContentInit, OnDestroy {
     } else {
       this.currentView = disks;
     }
+
+    // Setup event listener
+    if (this.views.length > 0) {
+      this.formEvent$ = new Subject<CoreEvent>();
+      this.formEvent$.pipe(
+        untilDestroyed(this),
```
Comment (undsoft) [inline src/app/pages/system/view-enclosure/view-enclosure.component.ts:267]: Looks like you're missing @UntilDestroy() decorator on component.

**Review by undsoft — APPROVED** [review]
Tested on a couple of random enclosures, including R20A. Things appear to be working correctly.

**Pull request merged**